File: test_sldr.m
function  Z = test_sldr(X, para)

% Z = test_sldr(X, para)
% Input:
%    X:      n x d matrix of original feature samples
%            d --- dimensionality of original features
%            n --- the number of samples
%    para:   structure of the model
% Output:
%    Z:      n x dim matrix of dimensionality reduced features

% Applies a trained SLDR (Supervised Linear Dimensionality Reduction) model to test data
% Inputs:
%   X    - (n x d) matrix of input samples
%   para - structure containing projection matrix and model parameters
% Output:
%   Z    - (n x dim) matrix of dimensionality-reduced features

% Retrieve the global mean from training data

mb = para.mb;

% recentering original feature
X = X - mb;

% Check which dimensionality reduction model was used
if  strcmp(para.model , 'plsda')
    % Partial Least Squares Discriminant Analysis path
    mu = para.mu_pca ;
    coeff = para.W;
    beta_coef = para.beta_coef;
    n = size(X,1);
    yhat = [ones(n,1) X]*beta_coef;
    Z = (yhat-mu)*coeff;
else
    W = para.W;
    Z = X*W;
end


% if strcmp(para.model , 'lda') || strcmp(para.model , 'hlda')
%
%     W = para.W;
%     mb = para.mb;
%     X = X - mb;
%     Z = X*W;
%
% elseif strcmp(para.model , 'mmda')
%
%     W = para.W;
%     mb = para.mb;
%     Sw_sqrtinv = para.Sw_sqrtinv;
%
%     X = X - mb;
%     X = X * Sw_sqrtinv ;
%     Z = X*W;
%
% end

File: test_sldr.py
import torch
import numpy as np


def test_sldr(X, para):
    """
    Apply a learned dimension reduction transformation to new data.
    
    Args:
        X: n x d tensor of original feature samples
            d --- dimensionality of original features
            n --- the number of samples
        para: dictionary of model parameters
    
    Returns:
        Z: n x dim tensor of dimensionality reduced features
    """
    # Convert to PyTorch tensor if not already
    if not isinstance(X, torch.Tensor):
        X = torch.tensor(X, dtype=torch.float32)
    
    # Get mean vector from parameters
    mb = para['mb']
    
    # Recenter original features
    X = X - mb
    
    # Apply the appropriate transformation based on model type
    if para['model'] == 'plsda':
        mu = para['mu_pca']
        coeff = para['W']
        beta_coef = para['beta_coef']
        
        n = X.shape[0]
        # Add column of ones for intercept term
        X_with_intercept = torch.cat([torch.ones(n, 1, dtype=X.dtype), X], dim=1)
        
        # Compute predictions
        yhat = X_with_intercept @ beta_coef
        
        # Apply PCA transformation
        Z = (yhat - mu) @ coeff
    else:
        # For LDA, HLDA, etc.
        W = para['W']
        Z = X @ W
    
    return Z

# The original code had commented sections for different model types, 
# but the active code only differentiated between 'plsda' and other models.
# If you need the functionality for 'mmda' or other specific model types in the 
# commented section, they can be implemented as additional conditions.